Facility config

jonboiser opened the pull request:
Implements Facility config page according to https://projects.invisionapp.com/share/QVB8RWZP2#/screens/228343981

![kapture 2017-04-14 at 17 26 12](https://cloud.githubusercontent.com/assets/10248067/25056540/937c2322-2137-11e7-9e75-7c8c76f2c1ee.gif)

## New Additions

* Add link to `managements/facilities` that goes to config page (although for only one facility ATM)
* Add actions related to facilityConfigPage in `faciltyConfigPageActions.js`
* Facility config page Vue files in '/views/facility-config-page'
* Added a separate file for mutations in `mutation.js` to facilitate unit-testing of mutations
* Initial implementation of a mocking framework for the 'kolibri' library in `kolibriGlobalMock.js`
* Unit tests for new actions, mutations, and views

## Changes to REST API

* Added `facility_id` filter to `datasets` endpoint

## Misc

* Fixed bug where `document.title` did not match `state.core.title` on initial load
* Some refactoring

indirectlylit commented on kolibri/core/assets/src/views/core-base.vue:
```diff
@@ -110,6 +115,7 @@
       },
     },
     mounted() {
+      this.updateDocumentTitle(this.title);
```
cool

rtibbles replied:
The other thing we need to do is get these titles into translatable strings - they are mostly not wrapped as of now.

indirectlylit replied:
Yeah - this relates to getting string wrapping in JS files, because titles are generated in actions. I'd started working on that a little but that got distracted.

rtibbles replied:
The wrapping is straight forward, we can easily invoke vue intl functionality anywhere, it's extracting the strings automatically (which is not perfect even in our current system) that's harder.

indirectlylit replied:
right sorry - that's what I meant. (At the time I was also looking at potentially auto-generating the namespaces based on file paths rather than having them hard-coded)

indirectlylit commented on kolibri/plugins/management/assets/src/state/facilityConfigPageActions.js:
```diff
@@ -0,0 +1,88 @@
+const { PageNames, defaultFacilityConfig, notificationTypes } = require('../constants');
+
+// When app is installed, the Facility is assigned id of `1`. Hardcoded here for now.
+const FACILITY_ID = 1;
```
Instead of hard-coding the facility ID, we might want to grab it from the user's session object?

The only caveat is that device superusers don't have an associated facility, which would limit this functionality facility admins only (which is also true of some other facility-specific functionality like coach tools)

indirectlylit commented on kolibri/plugins/management/assets/src/state/facilityConfigPageActions.js:
```diff
@@ -0,0 +1,88 @@
+  const ident = x => x;
+  return ConditionalPromise.all(promises)
+  .only(samePageCheckGenerator(store), ident, ident)._promise;
+}
```
This function is a little confusing to me - mind adding some comments about what it's doing?

Also, referencing `_promise` seems like a smell because typically `_` means internal-use. We may want to make it public, or create a getter?

jonboiser replied:
I've had to use the `_promise` escape hatch in a number of places because the chaining semantics of ConditionalPromise don't seem to match the Promise spec, esp. when it comes to error handling via `catch`.

I added some explanation in a comment, and maybe could look into extending ConditionalPromise later, although I'm more in favor of using something more standard like cancellable promises.

rtibbles replied:
I think we should move away from using ConditionalPromises by more judicious vuex state design. The main reason it is an issue now is because we are generally loading state for different pages into the same `pageState` top level. If we could partition and modularize our state better, this would be less of an issue and we could move away from my somewhat cludgey implementation of ConditionalPromises.

indirectlylit replied:
Cancellable promises also sound .. promising?

rtibbles replied:
Unfortunately, the javascript community did not agree: https://github.com/tc39/proposal-cancelable-promises

jonboiser replied:
One can get cancellable promises via a third-party library like Bluebird, which I think would make working with Promises a little more pleasant and consistent. But I think getting rid of the race condition that necessitates using ConditionalPromise.only so often would be even better. The redux middleware Redux-saga has some pretty sophisticated ways of handling races, I wonder if similar ideas are available in vuex. https://redux-saga.github.io/redux-saga/docs/advanced/RacingEffects.html

indirectlylit commented on kolibri/plugins/management/assets/src/state/facilityConfigPageActions.js:
```diff
@@ -0,0 +1,88 @@
+      facilityDatasetId: Number(dataset.id),
+      facilityName: facility.name,
+      // this part of state is mutated as user interacts with form
+      settings: sanitizeDataset(dataset),
```
In other places, we've used "API data to client state" mapping functions to explicitly map/pluck the values returned from the API to vuex state keys.

Even though simply dropping the contents of the API response into vuex is concise, I think it's a little messier long-term because:

 * there's no explicit place in js where the vuex state attributes get defined - instead it ends up implicitly defined in the vue files
 * server-side naming conventions (e.g. the underscores in `learner_can_edit_name`) start spilling into client-side code

indirectlylit commented on kolibri/plugins/management/assets/src/views/facilities-config-page/index.vue:
```diff
@@ -16,49 +16,16 @@
+          >
+            {{ $tr(setting) }}
+          </ui-checkbox>
+        </template>
```
any reason not to put the `v-for` directive directly on the `ui-checkbox`?

indirectlylit commented on kolibri/plugins/management/assets/src/views/facilities-config-page/index.vue:
```diff
@@ -113,6 +86,14 @@
+      learner_can_edit_name: 'Allow users to edit their full name',
+      learner_can_edit_password: 'Allow users to change their password when logged in',
+      learner_can_delete_account: 'Allow users to delete their account',
+      learner_can_sign_up: 'Allow users to sign-up on this device',
```
conventionally we use camelCase

indirectlylit commented:
usability-wise, one issue I noticed was that on smaller screens, the green 'save' feedback indicator is scrolled off-screen and not visible. (See below)

I guess this is why a [toaster-type pop-up](https://material.io/guidelines/components/snackbars-toasts.html) from the bottom can be useful. cc @khangmach @jtamiace - any thoughts on what pattern to use in these cases?


![image](https://cloud.githubusercontent.com/assets/2367265/25109927/397a3930-2395-11e7-884c-901ec4c038cf.png)

indirectlylit commented on kolibri/plugins/management/assets/src/state/facilityConfigPageActions.js:
```diff
@@ -0,0 +1,88 @@
+    const dataset = facilityDataset[0]; // assumes for now is only one Facility being managed
+    store.dispatch('SET_PAGE_STATE', {
+      // comes over wire as number, but gets changed to string
+      facilityDatasetId: Number(dataset.id),
```
I think we may actually want to standardize on string-based IDs. Eventually this may change to be a guid

jonboiser commented:
Made the following changes after review

1. Changed how `settings` is read (converted to camelCase) and saved (converted to snake_case)
2. Moved notifications to bottom of screen so it appears close to button.
3. Added some comments about ConditionalPromise wrapper `resolveOnlyIfOnSamePage`.
4. Grab `facility_id` from `session` part of store

khangmach commented:
@indirectlylit @jonboiser  ideally the success alert should be responsive to the screen. The main reason we're using keenUI alerts over snackbars in this case, is that because saving your facility settings can be seen as committing to one big operation (checking a bunch of stuff) and then hitting save. The user needs a visible and obvious indicator of success in this important case.

For the usage of snackbars, an example would be the actions checking different exercises to make an exam. We wouldn't incorporate success alerts for these smaller actions because the user would have to close the green alert every time something is added- which would be super obtrusive and counter-productive. Instead, we use small snackbars for quick feedback.

indirectlylit commented:
thanks Jonathan. Moving notifications to bottom seems like a reasonable quick-fix

> ideally the success alert should be responsive to the screen.

I see - in the mockups, [it's shown as a hovering bar](https://learningequality.gitbooks.io/design-iterations/content/learner-permission-configuration.html), not an inline bar.

I don't think we have a global component for that pattern but maybe we should make one and standardize on it.
